bwt/wavelet.go

```diff
@@ -0,0 +1,356 @@
+package bwt
+
+import (
+	"fmt"
+	"math"
+
+	"golang.org/x/exp/slices"
+)
+
+/*
+
+For the waveletTree's usage, please read the its
+method documentation. To understand what it is and how
+it works for either curiosity or maintenance, then read below.
+
+# WaveletTree
+
+The Wavelet Tree allows us to conduct RSA queries on strings. in
+a memory and run time efficient manner.
+RSA stands for (R)ank, (S)elect, (A)ccess.
+
+## The Character's Path Encoding
+
+One important component is a character's path encoding.
+Which character we are working with in a given path in the tree.
+For example, given the alphabet A B C D E F G, a possible encoding is:
+
+A: 000
+B: 001
+C: 010
+D: 011
+E: 100
+F: 101
+G: 110
+H: 111
+
+If we wanted to get to the leaf that represent the character D, we'd
+take the path:
+
+   root
+  /
+left
+  \
+ right
+    \
+   right
+
+## The Data Represented at each node
+
+Let us consider the sequence "bananas"
+It has the alphabet b, a, n, s
+Let's say it has the encoding:
+a: 00
+n: 01
+b: 10
+s: 11
+and that 0 if left and 1 is right
+We can represent this tree with bitvectors:
+
+     0010101
+    /       \
+  1000      001
+ /    \    /   \
+a      n  b     s
+
+If we translate each bit vector to its corresponding string, then it becomes:
+
+     bananas
+    /       \
+  baaa      nns
+ /    \    /   \
+a      b  n     s
+
+## RSA Intuition
+
+From here you may be able to build some intuition as to how we can take RSA queries given
+a characters path encoding and which character we'd like to Rank, Select, and Access.
+
+### Rank Example
+
+To get WaveletTree.Rank(a, 4) of bananas where a's encoding is 00
+1. root.Rank(0, 4) of 0010101 is 2
+2. Visit Left Child
+3. child.Rank(0, 2) of 1000 is 1
+4. Visit Left Child
+5. return 1
+
+### Select Example
+
+To get WaveletTree.Select(n, 1) of bananas where n's encoding is 01
+1. Go down to n's leaf using the path encoding is 01
+2. Go back to n's leaf's parent
+3. parent.Select(0, 1) of 001 is 1
+4. Go to the next parent
+5. parent.Select(1, 1) of 0010101 is 4
+6. return 4 since we are at the root.
+
+### Access Example
+
+If you've reached this point, then you must really be trying to understand how the
+waveletTree works. I recommend thinking through how access could work with the example
+above. HINT: rank might help.
```

Comment:
I think we should explain it explicity anyways for ease-of-use.

```suggestion
above. HINT: rank might help.

Take the tree we constructed earlier to represent the sequence "bananas".

     0010101
    /       \
  1000      001
 /    \    /   \
a      n  b     s

To access the 4th character of the sequence, we would call WaveletTree.Access(3),
which performs the following operations:
1. root[3] is 0 and root.Rank(0, 3) is 2
2. Since root[3] is 0, visit left child
3. child[2] is 0 and child.Rank(0, 2) is 1
4. Since child[2] is 0, visit left child
5. Left child is a leaf, so we've found our value (a)!

In pseudocode:

def f(node, index):
    if node is leaf:
        return character
    
    val := node[index]
    rank := node.Rank(val, index)

    nextNode := node.leftChild if val is 0, node.rightChild if val is 1
    return f(nextNode, rank)

```

Reply:
Fair. I'll definitely put in that explanation. 

The psuedocode could be redundant since the explanation is there along with the implementation below that could be used as a reference.